File: app/src/main/java/com/wa2c/android/medoly/plugin/action/lrclyrics/dialog/AboutDialogFragment.java
package com.wa2c.android.medoly.plugin.action.lrclyrics.dialog;

import android.app.AlertDialog;
import android.app.Dialog;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.text.Html;
import android.text.method.LinkMovementMethod;
import android.text.util.Linkify;
import android.util.TypedValue;
import android.view.Gravity;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.wa2c.android.medoly.plugin.action.lrclyrics.R;
import com.wa2c.android.medoly.plugin.action.lrclyrics.util.Logger;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * About dialog.
 */
public class AboutDialogFragment extends AbstractDialogFragment {

    /**
     * Create dialog instance.
     * @return Dialog instance.
     */
    static public AboutDialogFragment newInstance() {
        return new AboutDialogFragment();
    }



    /**
     * onCreateDialog
     */
    @SuppressWarnings("deprecation")
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        super.onCreateDialog(savedInstanceState);

        final View layoutView = View.inflate(getActivity(), R.layout.dialog_about, null);

        // Version
        try {
            PackageInfo packageInfo = getActivity().getPackageManager().getPackageInfo( getActivity().getPackageName(), PackageManager.GET_ACTIVITIES);
            ((TextView)layoutView.findViewById(R.id.dialogAboutAppVersionTextView)).setText(getString(R.string.label_dialog_about_ver, packageInfo.versionName));
        } catch (PackageManager.NameNotFoundException e) {
            Logger.e(e);
        }

        // Email
        ((TextView)layoutView.findViewById(R.id.dialogAboutEmailTextView)).setText(getString(R.string.app_mail_name) + "@" + getString(R.string.app_mail_domain));

        // License
        Linkify.addLinks(
                (TextView)layoutView.findViewById(R.id.dialogAboutLicenseTextView),
                Pattern.compile(getString(R.string.app_license)),
                getString(R.string.app_license_url),
                null,
                new Linkify.TransformFilter() {
                    @Override
                    public String transformUrl(Matcher match, String url) {
                        return getString(R.string.app_license_url);
                    }
                });

        // Privacy Policy
        Linkify.addLinks(
                (TextView)layoutView.findViewById(R.id.dialogAboutPrivacyPolicyTextView),
                Pattern.compile(getString(R.string.label_dialog_about_link)),
                getString(R.string.app_privacy_policy_url),
                null,
                new Linkify.TransformFilter() {
                    @Override
                    public String transformUrl(Matcher match, String url) {
                        return getString(R.string.app_privacy_policy_url);
                    }
                });

        // Google Play
        Linkify.addLinks(
                (TextView)layoutView.findViewById(R.id.dialogAboutGooglePlayTextView),
                Pattern.compile(getString(R.string.label_dialog_about_link)),
                getString(R.string.app_market_web),
                null,
                new Linkify.TransformFilter() {
                    @Override
                    public String transformUrl(Matcher match, String url) {
                        return getString(R.string.app_market_web);
                    }
                });


        // Library
        String[] libraryNames = getResources().getStringArray(R.array.about_library_names);
        String[] libraryUrls = getResources().getStringArray(R.array.about_library_urls);
        LinearLayout libraryLayout = (LinearLayout) layoutView.findViewById(R.id.dialogAboutLibraryLayout);
        for (int i = 0; i < libraryNames.length; i++) {
            TextView libTextView;
            libTextView = new TextView(getActivity());
            libTextView.setMovementMethod(LinkMovementMethod.getInstance());
            if (Build.VERSION.SDK_INT >= 24) {
                libTextView.setText(Html.fromHtml("<a href=\"" + libraryUrls[i] + "\">" +  libraryNames[i] + "</a>", Html.FROM_HTML_MODE_COMPACT));
            } else {
                libTextView.setText(Html.fromHtml("<a href=\"" + libraryUrls[i] + "\">" +  libraryNames[i] + "</a>"));
            }
            libTextView.setGravity(Gravity.CENTER_HORIZONTAL);
            libTextView.setTextSize(TypedValue.COMPLEX_UNIT_SP, 20);
            libraryLayout.setPadding(2, 2, 2, 2);
            libraryLayout.addView(libTextView);
        }

        // ダイアログ作成
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());
        builder.setTitle(R.string.pref_title_about);
        builder.setView(layoutView);
        builder.setPositiveButton(android.R.string.ok, null);
        return  builder.create();
    }

}

File: app/src/main/java/com/wa2c/android/medoly/plugin/action/lrclyrics/dialog/AboutDialogFragment.kt
package com.wa2c.android.medoly.plugin.action.lrclyrics.dialog

import android.app.Dialog
import android.content.pm.PackageManager
import android.os.Bundle
import android.text.method.LinkMovementMethod
import android.text.util.Linkify
import android.util.TypedValue
import android.view.Gravity
import android.view.LayoutInflater
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.core.text.HtmlCompat
import androidx.databinding.DataBindingUtil
import com.wa2c.android.medoly.plugin.action.lrclyrics.R
import com.wa2c.android.medoly.plugin.action.lrclyrics.databinding.DialogAboutBinding
import timber.log.Timber
import java.util.regex.Pattern

/**
 * About dialog.
 */
class AboutDialogFragment : AbstractDialogFragment() {

    /** Binding. */
    private lateinit var binding: DialogAboutBinding

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        super.onCreateDialog(savedInstanceState)
        binding = DataBindingUtil.inflate(LayoutInflater.from(context), R.layout.dialog_about, null, false)

        // Version
        try {
            val packageInfo = context.packageManager.getPackageInfo(context.packageName, PackageManager.GET_ACTIVITIES)
            binding.dialogAboutAppVersionTextView.text = getString(R.string.label_dialog_about_ver, packageInfo.versionName)
        } catch (e: PackageManager.NameNotFoundException) {
            Timber.e(e)
        }

        // Email
        val email = getString(R.string.app_mail_name) + "@" + getString(R.string.app_mail_domain)
        binding.dialogAboutEmailTextView.text = email

        // License
        Linkify.addLinks(
                binding.dialogAboutLicenseTextView,
                Pattern.compile(getString(R.string.app_license)),
                getString(R.string.app_license_url), null,
                Linkify.TransformFilter { _, _ -> getString(R.string.app_license_url) })

        // Privacy Policy
        Linkify.addLinks(
                binding.dialogAboutPrivacyPolicyTextView,
                Pattern.compile(getString(R.string.label_dialog_about_link)),
                getString(R.string.app_privacy_policy_url), null,
                Linkify.TransformFilter { _, _ -> getString(R.string.app_privacy_policy_url) })

        // Google Play
        Linkify.addLinks(
                binding.dialogAboutGooglePlayTextView,
                Pattern.compile(getString(R.string.label_dialog_about_link)),
                getString(R.string.app_market_web), null,
                Linkify.TransformFilter { _, _ -> getString(R.string.app_market_web) })

        // Library
        val libraryNames = resources.getStringArray(R.array.about_library_names)
        val libraryUrls = resources.getStringArray(R.array.about_library_urls)
        for (i in libraryNames.indices) {
            val libTextView = TextView(activity).apply {
                movementMethod = LinkMovementMethod.getInstance()
                text = HtmlCompat.fromHtml("<a href=\"" + libraryUrls[i] + "\">" + libraryNames[i] + "</a>", HtmlCompat.FROM_HTML_MODE_COMPACT)
                gravity = Gravity.CENTER_HORIZONTAL
                setTextSize(TypedValue.COMPLEX_UNIT_SP, 20f)
            }
            binding.dialogAboutLibraryLayout.setPadding(2, 2, 2, 2)
            binding.dialogAboutLibraryLayout.addView(libTextView)
        }

        // Build
        return AlertDialog.Builder(context).apply {
            setTitle(R.string.pref_title_about)
            setView(binding.root)
            setNeutralButton(android.R.string.ok, null)
        }.create()
    }

    companion object {
        /**
         * Create dialog instance.
         * @return Dialog instance.
         */
        fun newInstance(): AboutDialogFragment {
            return AboutDialogFragment().apply {
                arguments = Bundle()
            }
        }
    }

}
